File: frontend/src/util/glyph.ts
/*
 *     SPDX-License-Identifier: AGPL-3.0-only
 *
 *     Copyright (C) RainbowDashLabs and Contributor
 */
/**
 * The picture a row of gear is drawn with, resolved and ready to render.
 *
 * @property icon  the FontAwesome prefix and name, always present
 * @property color the colour it is drawn in, or null for the muted neutral
 */
export interface Glyph {
    icon: [string, string]
    color: string | null
}

/**
 * What is known about a row of gear when the picture is resolved.
 *
 * <p>Every field is optional, because the three payload shapes that feed this know different
 * amounts: a kind knows its own picture, a piece knows its kind's and its inventory's, and a
 * flattened row already carries the answer and resolves to itself.
 *
 * @property icon            a picture already resolved, or the row's own
 * @property color           the colour already resolved, or the row's own
 * @property artIcon         the picture of the piece's kind
 * @property artColor        the colour of the piece's kind
 * @property inventoryIcon   the picture of the inventory the piece is out of
 * @property inventoryColor  the colour of that inventory
 * @property homogeneous     whether that inventory is a stock rather than a collection, which decides
 *                           the shape when nothing else answers
 */
export interface GlyphSource {
    icon?: string | null
    color?: string | null
    artIcon?: string | null
    artColor?: string | null
    inventoryIcon?: string | null
    inventoryColor?: string | null
    homogeneous?: boolean
}

/** The shape a stock falls back to: one thing in many copies. */
const STOCK_FALLBACK = 'cube'

/** The shape a collection falls back to: a drawer of different things. */
const COLLECTION_FALLBACK = 'box'

function firstNamed(...candidates: Array<string | null | undefined>): string | null {
    for (const candidate of candidates) {
        if (candidate && candidate.trim()) return candidate.trim()
    }
    return null
}

/**
 * The one answer to "what is this drawn with", so no two screens disagree.
 *
 * <p>The kind answers first, then the inventory, then a plain shape. The two halves are resolved
 * together rather than one field at a time: a kind that set a colour and no picture is drawn in its
 * colour on the inventory's shape, which is what somebody choosing only a colour meant.
 *
 * @param source what is known about the row
 */
export function glyphFor(source: GlyphSource): Glyph {
    const icon = firstNamed(source.icon, source.artIcon, source.inventoryIcon)
        ?? (source.homogeneous === false ? COLLECTION_FALLBACK : STOCK_FALLBACK)
    const color = firstNamed(source.color, source.artColor, source.inventoryColor)
    return {icon: ['fas', icon], color}
}

File: src/main/java/dev/chojo/ember/feature/inventory/entity/Glyph.java
/*
 *     SPDX-License-Identifier: AGPL-3.0-only
 *
 *     Copyright (C) RainbowDashLabs and Contributor
 */
package dev.chojo.ember.feature.inventory.entity;

import io.javalin.http.BadRequestResponse;

import java.util.Locale;
import java.util.regex.Pattern;

/**
 * The picture a row of gear is drawn with: a FontAwesome name and the colour it is drawn in.
 *
 * <p>Both halves are optional and absent is the ordinary state. A reader falls back from the kind to
 * the inventory to a plain shape, so the pair travels together rather than as two fields that can be
 * half answered.
 *
 * <p>The name is not checked against anything. The catalogue of offerable pictures belongs to the
 * frontend, and a name it has retired must not make a save fail years later. The colour is checked,
 * because an unparseable colour would be painted.
 *
 * @param icon  the FontAwesome name, or {@code null} when nobody chose one
 * @param color the colour as {@code #rrggbb} in lower case, or {@code null} for the muted neutral
 */
public record Glyph(String icon, String color) {
    /**
     * Nothing chosen, which is what every row looks like before anybody picks a picture.
     */
    public static final Glyph NONE = new Glyph(null, null);

    private static final Pattern HEX_COLOR = Pattern.compile("^#[0-9a-fA-F]{6}$");

    /**
     * A glyph from what somebody sent, with blanks read as absent and the colour lowered.
     *
     * <p>A blank string and a missing field mean the same thing to a reader, and leaving both forms in
     * the column would make every comparison ask about whitespace.
     *
     * @param icon  the FontAwesome name, or {@code null}
     * @param color the colour, or {@code null}
     * @return the pair, with empty halves as {@code null}
     */
    public static Glyph of(String icon, String color) {
        return new Glyph(trimmedOrNull(icon), lowered(trimmedOrNull(color)));
    }

    /**
     * Whether the colour is one that can be painted, an absent colour included.
     *
     * @return {@code false} only for a colour that is present and not {@code #rrggbb}
     */
    public boolean validColor() {
        return color == null || HEX_COLOR.matcher(color).matches();
    }

    /**
     * Refuses a colour nothing could paint, which is the one half of a glyph worth checking.
     *
     * @throws BadRequestResponse when a colour is present and is not {@code #rrggbb}
     */
    public void requirePaintable() {
        if (!validColor()) {
            throw new BadRequestResponse("A colour is written as #rrggbb, and '%s' is not".formatted(color));
        }
    }

    /**
     * The same glyph in the one spelling a column keeps, refusing a colour nothing could paint.
     *
     * <p>What reaches a service has been built by a caller rather than by {@link #of}, so it is
     * normalised here as well. Two rows holding {@code #2563EB} and {@code #2563eb} would be two
     * colours to everything that compares them and one colour to everybody looking at them.
     *
     * @return the normalised pair
     * @throws BadRequestResponse when a colour is present and is not {@code #rrggbb}
     */
    public Glyph paintable() {
        Glyph normalised = of(icon, color);
        normalised.requirePaintable();
        return normalised;
    }

    private static String trimmedOrNull(String value) {
        if (value == null) return null;
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }

    private static String lowered(String value) {
        return value == null ? null : value.toLowerCase(Locale.ROOT);
    }
}
